core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -364,25 +344,67 @@ public void registerModelParser(MovAvgModel.AbstractModelParser parser) {
     }
 
     /**
-     * Enabling extending the get module by adding a custom aggregation parser.
+     * Register an aggregation.
+     *
+     * @param reader reads the aggregation builder from a stream
+     * @param aggregationParser reads the aggregation builder from XContent
+     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
+     *        is registered under.
+     */
+    public <AB extends AggregatorBuilder<AB>> void registerAggregation(Writeable.Reader<AB> reader, Aggregator.Parser aggregationParser,
+            ParseField aggregationName) {
+        aggregationParserRegistry.register(aggregationParser, aggregationName);
+        namedWriteableRegistry.register(AggregatorBuilder.class, aggregationName.getPreferredName(), reader);
+    }
+
+    /**
+     * Register an aggregation.
      *
-     * @param parser The parser for the custom aggregator.
+     * @deprecated prefer {@link #registerPipelineAggregation(Writeable.Reader, PipelineAggregator.Parser, ParseField)}. Will be removed
+     *             before 5.0.0GA.
      */
+    @Deprecated
     public void registerAggregatorParser(Aggregator.Parser parser) {
-        aggParsers.add(parser);
+        aggregationParserRegistry.register(parser, new ParseField(parser.type()));
+        namedWriteableRegistry.registerPrototype(AggregatorBuilder.class, parser.getFactoryPrototypes());
     }
 
+    /**
+     * Register a pipeline aggregation.
+     *
+     * @param reader reads the aggregation builder from a stream
+     * @param aggregationParser reads the aggregation builder from XContent
+     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
+     *        is registered under.
+     */
+    public <AB extends PipelineAggregatorBuilder<AB>> void registerPipelineAggregation(Writeable.Reader<AB> reader,
+            PipelineAggregator.Parser aggregationParser, ParseField aggregationName) {
+        pipelineAggregationParserRegistry.register(aggregationParser, aggregationName);
+        namedWriteableRegistry.register(PipelineAggregatorBuilder.class, aggregationName.getPreferredName(), reader);
+    }
+
+    /**
+     * Register a pipeline aggregation.
+     *
+     * @deprecated prefer {@link #registerPipelineAggregation(Writeable.Reader, PipelineAggregator.Parser, ParseField)}. Will be removed
+     *             before 5.0.0GA.
+     */
+    @Deprecated
```

Comment:
Is this only here until the rest of the aggregations are refactored? If so, could we add a NORELEASE comment here so we ensure its gone before 5.0 GA?